File: bricks/pac_native_plugin/__brick__/packages/{{name.snakeCase()}}/ios/Classes/Presentation/MviViewModel.swift
import Combine
import SwiftUI

public protocol BaseAction {}
public protocol BaseState {}
public protocol BaseEvent {}

open class MviViewModel<A: BaseAction, S: BaseState, E: BaseEvent>: ObservableObject {
    @Published public private(set) var state: S
    private let eventSubject = PassthroughSubject<E, Never>()
    public var events: AnyPublisher<E, Never> {
        eventSubject.eraseToAnyPublisher()
    }

    public init(initialState: S) {
        self.state = initialState
    }

    public func setState(_ update: (inout S) -> Void) {
        var newState = self.state
        update(&newState)
        self.state = newState
    }

    public func sendEvent(_ event: E) {
        eventSubject.send(event)
    }

    open func onAction(_ action: A) {
        // Override in subclass
    }
}

File: bricks/pac_add_native_ui/__brick__/packages/{{name.snakeCase()}}/android/src/main/kotlin/com/danhdue/{{name.snakeCase()}}/presentation/MviViewModel.kt
package com.danhdue.{{name.snakeCase()}}.presentation

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow

interface BaseAction
interface BaseState
interface BaseEvent

abstract class MviViewModel<A : BaseAction, S : BaseState, E : BaseEvent>(
    initialState: S
) {
    private val _state = MutableStateFlow(initialState)
    val state: StateFlow<S> = _state.asStateFlow()

    private val _event = Channel<E>(Channel.BUFFERED)
    val event = _event.receiveAsFlow()

    protected fun setState(update: S.() -> S) {
        _state.value = _state.value.update()
    }

    protected suspend fun sendEvent(event: E) {
        _event.send(event)
    }

    abstract fun onAction(action: A)
}
